soc/board_specific_workflows/test_general.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/env python3
+# Copyright 2021 The CFU-Playground Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import general
+import unittest
+import workflow_args
+from unittest import mock
+from litex.build import generic_platform, generic_programmer
+from litex.soc.integration import builder
+from litex.soc.integration import soc as litex_soc
+
+
+class TestGeneralSoCWorkflow(unittest.TestCase):
+    """TestCase for the GeneralSoCWorkflow class.
+    
+    All methods except 'run' are tested here.
+    """
+    def setUp(self):
+        """Sets up mocks and fakes used throughout the tests."""
+        # Generates a real argparse Namespace for testing.
+        self.target = 'test_target'
+        self.build = True
+        self.args = workflow_args.parse_workflow_args(
+            ['--target', self.target, '--build' if self.build else ''])
+
+        # Autospec the SoC.
+        self.programmer = mock.create_autospec(
+            generic_programmer.GenericProgrammer)
+        self.soc = mock.create_autospec(litex_soc.LiteXSoC, instance=True)
+        self.soc.platform = mock.create_autospec(
+            generic_platform.GenericPlatform, instance=True)
+        self.soc.platform.create_programmer.return_value = self.programmer
+        self.build_name = 'test_build_name'
+        self.soc.build_name = self.build_name
+
+        # Autospec the Builder.
+        self.builder = mock.create_autospec(builder.Builder, instance=True)
+        self.gateware_dir = 'test_gateware_dir/'
+        self.builder.gateware_dir = self.gateware_dir
+
+        self.builder_constructor = mock.create_autospec(
+            builder.Builder, return_value=self.builder)
+        # Can't easily autospec the soc_constructor -- interface inconsistent.
+        self.soc_constructor = mock.MagicMock(return_value=self.soc)
+
+    def simple_init(self):
+        """Returns a GeneralSoCWorkflow with reasonable testing parameters."""
+        return general.GeneralSoCWorkflow(self.args, self.soc_constructor,
+                                          self.builder_constructor)
+
+    def test_init(self):
+        """Tests functionality of __init__ method."""
```

Comment:
This test is not testing that the class is meeting its contract. Rather, it's testing implementation details. I would remove this test, in favor of tests that check that the externally observable behavior of the class is correct.

Reply:
That's a good point, this test has been removed.